pkg/balancer/server.go

```diff
@@ -0,0 +1,82 @@
+package balancer
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/util"
+)
+
+type Handler func(outbound net.Conn) error
+
+type Strategy interface {
+	// Route is expected to synchronously call the given Handle with an outbound connection.
+	Route(fn Handler) error
+	AddAddress(addr string) error
+	RemoveAddress(addr string) error
+	WaitRoutable(wait time.Duration) error
+}
+
+type Monitor interface {
+	MonitorHosts(service string, strategy Strategy, quitCh <-chan struct{})
+}
+
+type server struct {
+	logger  *logging.Logger
+	monitor Monitor
+}
+
+func DuplexConnection(inbound net.Conn) Handler {
+	return func(outbound net.Conn) error {
+		done := make(chan struct{})
+		defer inbound.Close()
+		defer outbound.Close()
```

Comment:
You're already closing `inbound` in `server.handleConnection()` and `outbound` in `LeastConnections.Route()`